servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/BeforeFinallyHttpOperator.java

```diff
@@ -142,44 +166,209 @@ public void request(final long n) {
 
                                             @Override
                                             public void cancel() {
-                                                try {
-                                                    if (stateUpdater.compareAndSet(ResponseCompletionSubscriber.this,
-                                                            PROCESSING_PAYLOAD, TERMINATED)) {
-                                                        beforeFinally.cancel();
+                                                if (!discardEventsAfterCancel) {
+                                                    try {
+                                                        if (stateUpdater.compareAndSet(
+                                                                ResponseCompletionSubscriber.this,
+                                                                PROCESSING_PAYLOAD, TERMINATED)) {
+                                                            beforeFinally.cancel();
+                                                        }
+                                                    } finally {
+                                                        subscription.cancel();
+                                                    }
+                                                    return;
+                                                }
+
+                                                for (;;) {
+                                                    final int state = ResponseCompletionSubscriber.this.state;
+                                                    assert state != IDLE;
+                                                    if (state == PROCESSING_PAYLOAD) {
```

Comment:
Could use a switch(state) but then you would have to use named break to distinguish between case/break and for/break

Reply:
Tried to be consistent with pattern we use at any other operator